### PR TITLE
Fix small performance regression reading None / NaN columns after Folly upgrade

### PR DESCRIPTION
**Pull request by poodlewars**
This fixes a small performance regression when reading columns with None or NaN values. We expect the change to only be measurable on very fast storage backends like LMDB or in-memory.

We could alternatively (mostly) fix this by just moving up to latest Folly - I'm hesitant to do that here as this saga has shown we are quite sensitive to Folly changes and it needs to be done carefully.

## Background

A user reported a slowdown between ArcticDB 5.10.0 and latest ArcticDB running reads like those in the `nonreg_random_read.py` benchmark. Their numbers were:

```
| version | 1M mean | 10M mean |
|---------|---------|----------|
| 5.10    | 0.060s  | 0.631s   |
| 6.12.1  | 0.077s  | 0.870s   |
```

A git bisect pinned the regression down to:

```
    38909e54a  Upgrade folly (#2502)      — folly 2023.09.25.00 → 2025.04.14.00
```

he issue only repro'd with the string and `col_none` columns. This narrowed the problem down to our `none_refcount_` and `nan_refcount_` handling in `python_handler_data.cpp`.

I then did another bisect within those Folly versions which pinned down the regression to [this change](https://github.com/facebook/folly/commit/2a3bd4c9b617cb6ff1f1331e6751f8b21d621193) in Folly which added extra synchronisation in `folly::ThreadLocal`. We acquire this new `forkHandlerLock_` whenever:
- A worker thread touches the None or NaN counts for the first time
- Three times when the `PythonHandlerData` is torn down

On my machine I got the following numbers, upgrading to latest Folly also helps us recover from the problem:

```
| build                                     | 1M mean (s) | 10M mean (s) |
|-------------------------------------------|-------------|--------------|
| 5.10.0 (folly 2023.09.25.00)              | 0.0484      | 0.447        |
| 6.13.0 (folly 2025.04.14.00)              | 0.0646      | 0.606        |
| master @ tree-two + folly 2026.02.23.00   | 0.0550      | 0.520        |
```

## Fix

I first added the `nonreg_random_read.py` ASV benchmark that repros the user's problem.

I also ran tests with a dataframe of 100k rows and 10k columns, with half the columns full of NaN and half full of None. These are the ones with `100kx10k` in the snippets below.

I simply changed the Folly thread locals for `std::atomic<uint64_t>` and this recovered the performance.

Without the fix:

```
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_1M
symbol: sym_1M  size: 1M
threads: cpu=16 io=24  warmup=2  iterations=20
mean=0.0635s  stdev=0.0034s  p50=0.0630s  min=0.0593s  max=0.0747s  n=20
```

```
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_100kx10k
symbol: sym_100kx10k  size-tag: 100kx10k
threads: cpu=16 io=24  warmup=2  iterations=20
mean=1.4774s  stdev=0.2175s  p50=1.5144s  min=0.9064s  max=1.7331s  n=20
```

With the move to `std::atomic`:

```
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_1M
symbol: sym_1M  size: 1M
threads: cpu=16 io=24  warmup=2  iterations=20
mean=0.0511s  stdev=0.0039s  p50=0.0504s  min=0.0451s  max=0.0633s  n=20
```

```
$ python ./bench_nan_none_read.py 
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_100kx10k
symbol: sym_100kx10k  size-tag: 100kx10k
threads: cpu=16 io=24  warmup=2  iterations=20
mean=1.3767s  stdev=0.2342s  p50=1.4399s  min=0.6685s  max=1.7649s  n=20
```

so we see in the narrow cases an improvement, and in the wide case the same results within stddev.

With a move to latest Folly instead:

```
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_1M
symbol: sym_1M  size: 1M
threads: cpu=16 io=24  warmup=2  iterations=20
mean=0.0497s  stdev=0.0029s  p50=0.0491s  min=0.0458s  max=0.0578s  n=20

python ./bench_nan_none_read.py 
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_100kx10k
symbol: sym_100kx10k  size-tag: 100kx10k
threads: cpu=16 io=24  warmup=2  iterations=20
mean=1.3818s  stdev=0.2577s  p50=1.3941s  min=0.4492s  max=1.6414s  n=20
```

With the old Folly version:

```
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_1M
symbol: sym_1M  size: 1M
threads: cpu=16 io=24  warmup=2  iterations=20
mean=0.0517s  stdev=0.0039s  p50=0.0513s  min=0.0473s  max=0.0634s  n=20
```

```
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_100kx10k
symbol: sym_100kx10k  size-tag: 100kx10k
threads: cpu=16 io=24  warmup=2  iterations=20
mean=1.4966 ...
```

*With 5.10.0+man4 we were still faster - this is due to the malloc_trim issues we discussed recently*

```
arcticdb version: 5.10.0+man4
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_100kx10k
symbol: sym_100kx10k  size-tag: 100kx10k
threads: cpu=16 io=24  warmup=2  iterations=20
mean=0.3864s  stdev=0.2791s  p50=0.3042s  min=0.2522s  max=1.2016s  n=20
```

We could fix this by just moving up to latest Folly - I'm hesitant to do that here as this saga has shown we are quite sensitive to Folly changes and it needs to be done carefully.

## Other Changes

https://peps.python.org/pep-0683/ introduces "immortal objects" (including `None`) from Python 3.12. Since we have shown here that the ref counting can be expensive, I've gotten rid of the `None` ref-counting when running on these Python versions.

We could set the `nan` as immortal too - but there is no public API in Python to do this so I chose not to.

I also made a small change to set the `NaN` refcount with one API call rather than repeated increments. This made no measurable difference to performance.

With Python 3.14 (so picking up the None change), with a dataframe of 50% None cols and 50% NaN cols:

```
(tree-two-py314) aseaton@dlonapatcs502:~/source/ArcticDB-worktree/tree-two/python DEV $ python ./bench_nan_none_read.py 
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_100kx10k
symbol: sym_100kx10k  size-tag: 100kx10k
threads: cpu=16 io=24  warmup=2  iterations=20
mean=1.5837s  stdev=0.2243s  p50=1.6384s  min=0.9165s  max=1.8272s  n=20
```

whereas without these changes we had the same within stddev:

```
$ python ./bench_nan_none_read.py 
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_100kx10k
symbol: sym_100kx10k  size-tag: 100kx10k
threads: cpu=16 io=24  warmup=2  iterations=20
mean=1.4971s  stdev=0.2179s  p50=1.5409s  min=0.7825s  max=1.8207s  n=20
```

on the normal data:

```
python ./bench_read.py 
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_1M
symbol: sym_1M  size: 1M
threads: cpu=16 io=24  warmup=2  iterations=20
mean=0.0545s  stdev=0.0040s  p50=0.0537s  min=0.0503s  max=0.0644s  n=20
```

whereas without these changes we had:

```
$ python ./bench_bbg_read.py 
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_1M
symbol: sym_1M  size: 1M
threads: cpu=16 io=24  warmup=2  iterations=20
mean=0.0634s  stdev=0.0046s  p50=0.0617s  min=0.0580s  max=0.0737s  n=20
```

### CODE REVIEW

File: python/benchmarks/nonreg_random_read.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+from asv_runner.benchmarks.mark import SkipNotImplemented
+
+from arcticdb.util.logger import get_logger
+from benchmarks.environment_setup import Storage, create_libraries
+
+SEED = 20260421
+STORAGES = [Storage.LMDB, Storage.AMAZON]
+
+
+def _make_strings(n, rng, length=8):
```
Comment (vasil-pashov) [inline python/benchmarks/nonreg_random_read.py:34]: We have a couple of utility functions to generate random strings here https://github.com/man-group/ArcticDB/blob/144e5f9c52d7e8fbbe0c1232ba027c8e35c96722/python/arcticdb/util/test.py#L282

**Review by vasil-pashov — APPROVED** [review]

**Comment by poodlewars** [comment]
Repeating the benchmarking,

 | Version | Shape | mean (s) | stdev (s) | p50 (s) | min (s) | max (s) |                                                                                                                                
  |---|---|---|---|---|---|---|                                                                                                                                                                           
  | 5.10.0 | narrow (1M) | 0.0471 | 0.0047 | 0.0464 | 0.0423 | 0.0618 |                                                                                                                                   
  | master (no fix) | narrow (1M) | 0.0624 | 0.0035 | 0.0619 | 0.0580 | 0.0733 |
  | master (std::atomic) | narrow (1M) | 0.0489 | 0.0041 | 0.0483 | 0.0444 | 0.0596 |                                                                                                                     
  | 5.10.0 | wide (100kx10k) | 0.3828 | 0.2704 | 0.3008 | 0.2503 | 1.1955 |                                                                                                                               
  | master (no fix) | wide (100kx10k) | 1.2470 | 0.1695 | 1.2389 | 0.9281 | 1.5305 |                                                                                                                      
  | master (std::atomic) | wide (100kx10k) | 1.2672 | 0.1998 | 1.2678 | 0.8754 | 1.5728 |                                                                                                                 

I also tried swapping out the counters to `oneapi::tbb::combinable` which is a competitor to `folly::ThreadCachedInt` with these results:

```
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_1M
symbol: sym_1M  size: 1M
threads: cpu=16 io=24  warmup=2  iterations=20
mean=0.0501s  stdev=0.0040s  p50=0.0496s  min=0.0448s  max=0.0616s  n=20

arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_10M
symbol: sym_10M  size: 10M
threads: cpu=16 io=24  warmup=2  iterations=20
mean=0.5399s  stdev=0.0106s  p50=0.5378s  min=0.5220s  max=0.5615s  n=20

(tree-two-venv) aseaton@dlonapatcs502:~/source/ArcticDB-worktree/tree-two/python DEV $ python ./bench_nan_none_read.py 
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_100kx10k
symbol: sym_100kx10k  size-tag: 100kx10k
threads: cpu=16 io=24  warmup=2  iterations=20
mean=1.3031s  stdev=0.2335s  p50=1.3798s  min=0.5157s  max=1.5342s  n=20
```

Finally I tried with 64 cpu threads and 96 IO threads, on a 10M row symbol,

```
sharded atomic counters
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_10M
symbol: sym_10M  size: 10M
threads: cpu=64 io=96  warmup=2  iterations=20
mean=0.5679s  stdev=0.0103s  p50=0.5651s  min=0.5539s  max=0.5938s  n=20

std::atomic
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_10M
symbol: sym_10M  size: 10M
threads: cpu=64 io=96  warmup=2  iterations=20
mean=0.5673s  stdev=0.0124s  p50=0.5665s  min=0.5543s  max=0.6111s  n=20

tbb
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_10M
symbol: sym_10M  size: 10M
threads: cpu=64 io=96  warmup=2  iterations=20
mean=0.5742s  stdev=0.0081s  p50=0.5727s  min=0.5600s  max=0.5943s  n=20

master
arcticdb version: dev
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_10M
symbol: sym_10M  size: 10M
threads: cpu=64 io=96  warmup=2  iterations=20
mean=0.6824s  stdev=0.0059s  p50=0.6814s  min=0.6673s  max=0.6923s  n=20

5.10
arcticdb version: 5.10.0
lmdb: /data/team/data/arctic_native/aseaton/bbg_random_20260421/lmdb_dev_10M
symbol: sym_10M  size: 10M
threads: cpu=64 io=96  warmup=2  iterations=20
mean=0.5319s  stdev=0.0087s  p50=0.5301s  min=0.5222s  max=0.5501s  n=20
```

**Review by alexowens90 — APPROVED** [review]
